src/storage/compressed_tile.cpp

```diff
@@ -0,0 +1,588 @@
+
+  base_value = column_values[num_tuples / 2];
+  type::Value maxDiff = column_values[num_tuples - 1].Subtract(base_value);
+  type::Value minDiff = column_values[0].Subtract(base_value);
+
+  std::vector<type::Value> modified_values(num_tuples);
+
+  while (true) {
+    type::Value maxValue = type::Type::GetMaxValue(compression_type);
+    type::Value minValue = type::Type::GetMinValue(compression_type);
+
+    if ((maxDiff.CompareLessThanEquals(maxValue) == type::CMP_TRUE) &&
+        (minDiff.CompareGreaterThanEquals(minValue) == type::CMP_TRUE)) {
+      // Found a  suitable compression type.
+      for (oid_t k = 0; k < num_tuples; k++) {
+        modified_values[k] =
+            actual_values[k].Subtract(base_value).CastAs(compression_type);
+      }
+      break;
+    } else {
+      if (type::Type::GetTypeSize(column_type) ==
+          type::Type::GetTypeSize(static_cast<type::Type::TypeId>(
+              static_cast<int>(compression_type) + 1))) {
+        modified_values.resize(0);
+        break;
+      }
+
+      compression_type = static_cast<type::Type::TypeId>(
+          static_cast<int>(compression_type) + 1);
+    }
+  }
+  return modified_values;
+}
+
+// compression for varchar
+std::vector<type::Value> CompressedTile::CompressCharColumn(Tile *tile,
+                                                            oid_t column_id) {
+  oid_t num_tuples = tile->GetAllocatedTupleCount();
+  auto tile_schema = tile->GetSchema();
+  bool is_inlined = tile_schema->IsInlined(column_id);
+  size_t column_offset = tile_schema->GetOffset(column_id);
+  auto column_type = tile_schema->GetType(column_id);
+  std::vector<type::Value> column_values(num_tuples);
+
+  // use 3rd party
+  cuckoohash_map<std::string, int> dictionary;
+
+  std::vector<type::Value> decoder;
+
+  std::vector<type::Value> modified_values(num_tuples);
+
+  /* put data here first.
+   * since we do not know how many uniq words at the beginning
+   * first store them as int
+   * decide exact Value type later
+   */
+
+  std::vector<int> modified_raw(num_tuples);
+
+  int counter = 0;
+  int new_value = 0;
+  for (oid_t i = 0; i < num_tuples; i++) {
+    type::Value val =
+        tile->GetValueFast(i, column_offset, column_type, is_inlined);
+    std::string word = val.ToString();
+
+    if (dictionary.contains(word) == false) {
+      // add a new word to dictionary
+      dictionary.insert(word, counter);
+      decoder.push_back(type::ValueFactory::GetVarcharValue(word));
+      LOG_TRACE("new word: #%d : %s", counter,
+                decoder.at(counter).ToString().c_str());
+      new_value = counter;
+      counter++;
+
+    } else {
+      new_value = dictionary.find(word);
+    }
+    modified_raw[i] = new_value;
+  }
+
+  LOG_TRACE("number of uniq words: %d", counter);
+  if ((oid_t)counter == num_tuples) {
+    // no duplicate
+    LOG_TRACE("All words are unique");
+    modified_values.clear();
```

Comment:
It seems this is unnecessary, modified_values is not modified up to this point anyways.